aiidalab/app.py

```diff
@@ -0,0 +1,647 @@
+# -*- coding: utf-8 -*-
+"""Module to manage AiiDA lab apps."""
+
+import re
+import os
+from os import path
+from time import sleep
+from collections import OrderedDict
+
+import requests
+import ipywidgets as ipw
+from dulwich.repo import Repo
+
+
+class AppNotInstalledException(Exception):
+    pass
+
+
+class Version(ipw.VBox):
+    """Class to choose app's version."""
+
+    def __init__(self):
+        self.change_btn = ipw.Button(description="choose seleted")
+        self.selected = ipw.Select(
+            options={},
+            description='Select version',
+            disabled=False,
+            style={'description_width': 'initial'},
+        )
+        self.info = ipw.HTML('')
+        super().__init__([self.selected, self.change_btn, self.info])
+
+
+class AiidalabApp():  # pylint: disable=attribute-defined-outside-init,too-many-public-methods,too-many-instance-attributes
+    """Class to manage AiiDA lab app."""
+
+    def __init__(self, name, app_data, aiidalab_apps):  #, custom_update=False):
+        if app_data is not None:
+            self._git_url = app_data['git_url']
+            self._meta_url = app_data['meta_url']
+            self._git_remote_refs = app_data['gitinfo']
+            self.categories = app_data['categories']
+        else:
+            self._git_url = None
+            self._git_remote_refs = {}
+        self.install_info = ipw.HTML()
+        self.aiidalab_apps = aiidalab_apps
+        self.name = name
+
+    def in_category(self, category):
+        # One should test what happens if the category won't be defined.
+        return category in self.categories
+
+    def _get_appdir(self):
+        return path.join(self.aiidalab_apps, self.name)
+
+    def is_installed(self):
+        """The app is installed if the corresponding folder is present."""
+        return path.isdir(self._get_appdir())
+
+    def has_git_repo(self):
```

Comment:
I am likely going to propose to use a distinct class to manage "git-installed" apps, but I am ok for leaving these functions here for now as long as it is clear that they are likely going to be removed.

Reply:
As far as I know, they are used only internally for the moment. So it is fine.